### PR TITLE
Annoy/Usearch index: Fix LOGICAL_ERROR during build-up with default values

### LINKED ISSUE #52258: LOGICAL_ERROR in Annoy index: Array has [...] rows, [...] rows expected.
```
<Error> executeQuery: Code: 49. DB::Exception: Array has 0 rows, 20 rows expected. (LOGICAL_ERROR) (version 23.5.1.34446 (official build)) (from 10.20
.5.65:60124) (in query: [...] Stack trace (when copying this message, always include the lines below):

0. DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x000000000dd460d5 in /usr/bin/clickhouse
1. ? @ 0x00000000087a5e50 in /usr/bin/clickhouse
2. ? @ 0x00000000143cf0c8 in /usr/bin/clickhouse
3. DB::MergeTreeDataPartWriterOnDisk::calculateAndSerializeSkipIndices(DB::Block const&, std::vector<DB::Granule, std::allocator<DB::Granule>> const&) @ 0x0000000014343de6 in /usr/bin/clickhouse
4. DB::MergeTreeDataPartWriterCompact::writeDataBlockPrimaryIndexAndSkipIndices(DB::Block const&, std::vector<DB::Granule, std::allocator<DB::Granule>> const&) @ 0x0000000014335dfa in /usr/bin/clickhouse
5. DB::MergeTreeDataPartWriterCompact::fillDataChecksums(DB::MergeTreeDataPartChecksums&) @ 0x00000000143371ef in /usr/bin/clickhouse
6. DB::MergeTreeDataPartWriterCompact::fillChecksums(DB::MergeTreeDataPartChecksums&) @ 0x000000001433803c in /usr/bin/clickhouse
7. ? @ 0x00000000144732ca in /usr/bin/clickhouse
8. DB::MergeTreeDataWriter::writeTempPartImpl(DB::BlockWithPartition&, std::shared_ptr<DB::StorageInMemoryMetadata const> const&, std::shared_ptr<DB::Context const>, long, bool) @ 0x000000001446c3b4 in /usr/bin/clickhouse
9. DB::ReplicatedMergeTreeSinkImpl<false>::consume(DB::Chunk) @ 0x0000000014589794 in /usr/bin/clickhouse
10. DB::SinkToStorage::onConsume(DB::Chunk) @ 0x0000000014adeb06 in /usr/bin/clickhouse
11. ? @ 0x0000000014a25f6b in /usr/bin/clickhouse
12. ? @ 0x0000000014a25c9b in /usr/bin/clickhouse
13. DB::ExceptionKeepingTransform::work() @ 0x0000000014a25558 in /usr/bin/clickhouse
14. DB::ExecutionThreadContext::executeTask() @ 0x0000000014820d23 in /usr/bin/clickhouse
15. DB::PipelineExecutor::executeStepImpl(unsigned long, std::atomic<bool>*) @ 0x0000000014817070 in /usr/bin/clickhouse
16. ? @ 0x0000000014818b03 in /usr/bin/clickhouse
17. ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::worker(std::__list_iterator<ThreadFromGlobalPoolImpl<false>, void*>) @ 0x000000000de217c3 in /usr/bin/clickhouse
18. void std::__function::__policy_invoker<void ()>::__call_impl<std::__function::__default_alloc_func<ThreadFromGlobalPoolImpl<false>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleI
mpl<void>(std::function<void ()>, long, std::optional<unsigned long>, bool)::'lambda0'()>(void&&)::'lambda'(), void ()>>(std::__function::__policy_storage const*) @ 0x000000000de23fb5 in /usr/bin/clickhouse
19. ThreadPoolImpl<std::thread>::worker(std::__list_iterator<std::thread, void*>) @ 0x000000000de1d654 in /usr/bin/clickhouse
20. ? @ 0x000000000de22e41 in /usr/bin/clickhouse
21. ? @ 0x00007ff6ee17bb43 in ?
22. ? @ 0x00007ff6ee20da00 in ?

```

This issue happened during build-up of the [experimental Annoy index](https://clickhouse.com/docs/en/engines/table-engines/mergetree-family/annindexes#annoy) in an undisclosed production system because of which it cannot be easily be reproduced (I tried simpler queries but the error was not triggered). If someone hits the same issue and can share a repro, please post it here - thanks!

### PR DESCRIPTION
**Pull request by rschu1ze**
Fixes: #52258

### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix a LOGICAL_ERROR when empty or default arrays are inserted into Annoy and Usearch indexes

### CODE REVIEW

**Comment by robot-ch-test-poll3** [comment]
*This is an automated comment for commit fc5eb6606ff67355df8b8fb907a3f9220bcb85f6 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/54600/fc5eb6606ff67355df8b8fb907a3f9220bcb85f6/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>✅ success</td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>⏳ pending</td></tr>
<tbody>
</table>

**Comment by rschu1ze** [comment]
@devcrafter Like to check / approve? Thanks!

File: src/Storages/MergeTree/MergeTreeIndexAnnoy.cpp [hunk 1]
```diff
@@ -154,36 +154,43 @@ void MergeTreeIndexAggregatorAnnoy<Distance>::update(const Block & block, size_t
...
-        if (array.empty())
-            throw Exception(ErrorCodes::LOGICAL_ERROR, "Array has 0 rows, {} rows expected", rows_read);
+        const auto & column_array_offsets = column_array->getOffsets();
+        const size_t num_rows = column_array_offsets.size();
 
-        const auto & offsets = column_array->getOffsets();
-        const size_t num_rows = offsets.size();
+        /// The index dimension is inferred from the inserted arrays (array cardinality). If no value was specified in the INSERT statement
+        /// for the annoy-indexed column (i.e. default value), we have a problem. Reject such values.
+        if (column_array_offsets.empty() || column_array_offsets[0] == 0)
```
Comment (devcrafter) [inline src/Storages/MergeTree/MergeTreeIndexAnnoy.cpp:165]: Interesting why `column_array_offsets.empty()` is not enough? What special meaning does the condition`column_array_offsets[0] == 0` imply?

Comment (rschu1ze) [reply src/Storages/MergeTree/MergeTreeIndexAnnoy.cpp]: Thanks - the code now looks less weird.

File: src/Storages/MergeTree/MergeTreeIndexAnnoy.cpp [hunk 2]
```diff
@@ -154,36 +154,45 @@ void MergeTreeIndexAggregatorAnnoy<Distance>::update(const Block & block, size_t
 
     if (const auto & column_array = typeid_cast<const ColumnArray *>(column_cut.get()))
     {
-        const auto & data = column_array->getData();
-        const auto & array = typeid_cast<const ColumnFloat32 &>(data).getData();
+        const auto & column_array_data = column_array->getData();
+        const auto & column_arary_data_float_data = typeid_cast<const ColumnFloat32 &>(column_array_data).getData();
```
Comment (devcrafter) [inline src/Storages/MergeTree/MergeTreeIndexAnnoy.cpp:158]: So much data
```suggestion
        const auto & column_arary_data_float = typeid_cast<const ColumnFloat32 &>(column_array_data).getData();
```

File: src/Storages/MergeTree/MergeTreeIndexAnnoy.cpp [hunk 3]
```diff
@@ -154,36 +154,45 @@ void MergeTreeIndexAggregatorAnnoy<Distance>::update(const Block & block, size_t
...
+        const auto & column_array_offsets = column_array->getOffsets();
+        const size_t num_rows = column_array_offsets.size();
 
-        const auto & offsets = column_array->getOffsets();
-        const size_t num_rows = offsets.size();
+        /// The Annoy algorithm naturally assumes that the indexed vectors have dimension >= 1. This condition is violated if empty arrays
+        /// are INSERTed into an Annoy-indexed column or if no value was specified at all in which case the arrays take on their default
+        /// value which is also empty.
+        if (column_array->isDefaultAt(0))
+            throw Exception(ErrorCodes::INCORRECT_DATA, "The arrays in column '{}' must not be empty. Did you try to INSERT default values?", index_column_name);
```
Comment (devcrafter) [inline src/Storages/MergeTree/MergeTreeIndexAnnoy.cpp:167]: For a user, it's a present error, so probably `Did you try ...` should be in present time

File: src/Storages/MergeTree/MergeTreeIndexAnnoy.cpp [hunk 4]
```diff
@@ -154,36 +154,45 @@ void MergeTreeIndexAggregatorAnnoy<Distance>::update(const Block & block, size_t
...
-            throw Exception(ErrorCodes::LOGICAL_ERROR, "Array has 0 rows, {} rows expected", rows_read);
+        const auto & column_array_offsets = column_array->getOffsets();
+        const size_t num_rows = column_array_offsets.size();
 
-        const auto & offsets = column_array->getOffsets();
-        const size_t num_rows = offsets.size();
+        /// The Annoy algorithm naturally assumes that the indexed vectors have dimension >= 1. This condition is violated if empty arrays
+        /// are INSERTed into an Annoy-indexed column or if no value was specified at all in which case the arrays take on their default
+        /// value which is also empty.
+        if (column_array->isDefaultAt(0))
```
Comment (devcrafter) [inline src/Storages/MergeTree/MergeTreeIndexAnnoy.cpp:166]: Shouldn't we check the size of `offsets_data` first?

File: src/Storages/MergeTree/MergeTreeIndexAnnoy.cpp (hunk 2 above)
Comment (rschu1ze) [reply src/Storages/MergeTree/MergeTreeIndexAnnoy.cpp:158]: There is some logic behind the naming. Arrays and tuples are nested structures, so, yes, there will be a few `data` in variables.

File: src/Storages/MergeTree/MergeTreeIndexAnnoy.cpp (hunk 4 above)
Comment (rschu1ze) [reply src/Storages/MergeTree/MergeTreeIndexAnnoy.cpp:166]: This looks admittedly a bit silly. l. 143 has an early exit if `rows_read == 0`. There could theoretically the situation that `rows_read != 0` but the array is still empty (in which case we'd crash here). But in such a case, something is badly broken already anyways. I may add another check for that in a follow-up PR.

**Review by devcrafter — APPROVED** [review]

File: src/Storages/MergeTree/MergeTreeIndexAnnoy.cpp (hunk 4 above)
Comment (rschu1ze) [reply src/Storages/MergeTree/MergeTreeIndexAnnoy.cpp:166]: Totally, just in the next PR. Thanks!

**Pull request merged**